README.md
# ChatGPT Web

### 使用界面
![cover3](./docs/c3.png)
![cover](./docs/c1.png)
![cover2](./docs/c2.png)

### 性格微调功能
![cover4](./docs/c4.png)

### 文本审查功能(使用OpenAI官方接口)
![cover5](./docs/c5.png)

- [ChatGPT Web](#chatgpt-web)
	- [介绍](#介绍)
	- [快速部署](#快速部署)
	- [开发环境搭建](#开发环境搭建)
		- [Node](#Node)
		- [Python](#Python)
	- [开发环境启动项目](#开发环境启动项目)
		- [后端](#后端服务)
		- [前端](#前端网页)
	- [打包和部署](#打包和部署)
		- [前端打包](#前端资源打包(需要安装node和docker、docker-compose))
		- [后端打包](#后端服务打包)
	- [使用DockerCompose启动](#使用DockerCompose启动)
	- [常见问题](#常见问题)
	- [参与贡献](#参与贡献)
	- [赞助原作者](#赞助)
	- [License](#license)

## 介绍

这是一个可以自己在本地部署的`ChatGpt`网页，使用`OpenAI`的官方API接入`gpt-3.5-turbo`或`gpt-4`模型来实现接近`ChatGPT Plus`的对话效果。
源代码Fork和修改于[Chanzhaoyu/chatgpt-web](https://github.com/Chanzhaoyu/chatgpt-web/)

与OpenAI官方提供的付费版本`ChatGPT Plus`对比，`ChatGPT Web`有以下优势：

1. **省钱(仅限`gpt-3.5-turbo`)**。你可以用1折左右的价格，体验与`ChatGPT Plus`
	 几乎相同的对话服务。对于日常学习工作使用，你不必每月花费20$购买Plus服务，使用本项目自己对接API的使用成本，是`ChatGPT Plus`
	 的1/10左右。
2. **0门槛使用**。你可以将自建的`ChatGPT Web`
	 站点分享给家人和朋友，他们不再需要经历“解决网络问题”-“登录”-“输入密码”-“点选验证码”这一连串麻烦的环节，就可以轻松享受到`ChatGPT Plus`
	 带来的生产力提升。
3. **可以缓解网络封锁的影响**。`ChatGPT Web`只需要一个`OpenAI API Key`即可使用，如果你所在的地区无法访问`OpenAI`
	 ，你可以将`ChatGPT Web`部署在海外服务器上，或在当地服务器上配置`socks_proxy`参数来转发请求给代理软件，即可正常使用。

和[Chanzhaoyu的原版](https://github.com/Chanzhaoyu/chatgpt-web/)的主要区别：

1. 专注于易用、易部署、不操心，我将尽我所能做到对小白用户友好，因此本项目也会舍弃一些专业功能，例如：

	 - 不支持accessToken这类非官方使用方式。我认为你只是希望有个稳定能用的AI助手，并不想折腾这些
	 - 不支持反向代理。因为很多用户喜欢使用第三方给的反代地址，但是它们的安全性存疑、封号也超级快！
   - 不做导入和管理Prompt模板的功能。使用者没有“这个链接干什么用的”、“什么是json文件”、“为什么要我自己审核json文件的安全性，怎么审核？”此类烦恼

2. 可以识别语音消息：通过接入`whisper-1`API实现。懒得打字的时候很好用
3. 可以微调参数：通过设置可以调整`ChatGpt`的性格
4. 可以“记住”更多的上下文：聊天记录保存在服务器端，不会因为网页端清理缓存而丢失上下文，影响回答质量

其它区别：
1. 增加日语界面
2. 优化了移动端体验

## 快速部署

如果你不需要自己开发，只需要部署使用，可以直接跳到 [使用最新版本docker镜像启动](#使用最新版本docker镜像启动)

## 开发环境搭建

### Node

`node` 需要 `^16 || ^18` 版本（`node >= 14`
需要安装 [fetch polyfill](https://github.com/developit/unfetch#usage-as-a-polyfill)
），使用 [nvm](https://github.com/nvm-sh/nvm) 可管理本地多个 `node` 版本

```shell
node -v
```

如果你没有安装过 `pnpm`

```shell
npm install pnpm -g
```

### Python

`python` 需要 `3.10` 以上版本，进入文件夹 `/service` 运行以下命令

```shell
pip install --no-cache-dir -r requirements.txt
```

## 开发环境启动项目

### 启动前端服务

根目录下运行以下命令

```shell
# 前端网页的默认端口号是1002，对接的后端服务的默认端口号是3002，可以在 .env 和 .vite.config.ts 文件中修改
pnpm bootstrap
pnpm dev
```

### 启动后端服务

只有`--openai_api_key`是必填的启动参数，需要先去[OpenAI](https://platform.openai.com/)
注册账号，然后在[这里](https://platform.openai.com/account/api-keys)获取`OPENAI_API_KEY`。

```shell
# 进入文件夹 `/service` 运行以下命令
python main.py --openai_api_key="$OPENAI_API_KEY"
```

除了`openai_api_key`这个必填的参数之外，还有以下可选参数可用：

- `openai_timeout_ms` 访问OpenAI的超时时间(毫秒)，默认值为 '100000'
- `api_model` 默认值为 gpt-3.5-turbo 也可以使用很贵的 gpt-4
- `socks_proxy` 代理，默认值为空字符串，格式示例: `http://127.0.0.1:10808`
- `host` HOST，默认值为 0.0.0.0
- `port` PORT，默认值为 3002

也就是说你也可以这样启动
```shell
python main.py --openai_api_key="$OPENAI_API_KEY" --openai_timeout_ms="$OPENAI_TIMEOUT_MS" --api_model="$API_MODEL" --socks_proxy="$SOCKS_PROXY" --host="$HOST" --port="$PORT"
```


## 打包和部署

### 前端资源打包(需要安装node)

1. 根目录下运行以下命令
	 ```shell
	 pnpm run build
	 ```
2. 将打包好的文件夹`dist`文件夹复制到`/docker-compose/nginx`目录下，并改名为`html`
	```shell
	cp dist/ docker-compose/nginx/html -r
	```
3. 配置访问权限
	```shell
	  # 进入文件夹 `/docker-compose/nginx`
    cd docker-compose/nginx
    # 运行add_user.sh脚本，根据提示创建用户名和密码
    # (密码文件将被保存在 /docker-compose/nginx/auth/.htpasswd)
    bash add_user.sh
    # 如果你想删除一个用户，可以使用remove_user.sh脚本
    bash remove_user.sh
	```

### 后端服务打包为docker容器(需要安装docker和docker-compose)

1. 进入文件夹 `/service` 运行以下命令
	 ```shell
	 docker build -t chatgpt-web-backend .
	 ```

### 使用DockerCompose启动

- 进入文件夹 `/docker-compose` 修改 `docker-compose.yml` 文件

	```
  version: '3'

  services:
    app:
      image: chatgpt-web-backend # 这里填你打包的后端服务的镜像名字
      ports:
        - 3002:3002
      environment:
        OPENAI_API_KEY: your_openai_api_key
        # 访问OpenAI的超时时间(毫秒)，可选，默认值为 '100000'
        OPENAI_TIMEOUT_MS: '100000'
        # 可选，默认值为 gpt-3.5-turbo
        API_MODEL: gpt-3.5-turbo
        # Socks代理，可选，格式为 http://127.0.0.1:10808
        SOCKS_PROXY: ''
        # HOST，可选，默认值为 0.0.0.0
        HOST: 0.0.0.0
        # PORT，可选，默认值为 3002
        PORT: 3002
    nginx:
      depends_on:
        - app
      image: nginx:alpine
      ports:
        - '80:80'
      expose:
        - '80'
      volumes:
        - ./nginx/html/:/etc/nginx/html/
        - ./nginx/auth/:/etc/nginx/auth/
        - ./nginx/nginx.conf:/etc/nginx/conf.d/default.conf
      links:
        - app
	```

- 进入文件夹 `/docker-compose` 运行以下命令

	```shell
	# 前台运行
	docker-compose up
	# 或后台运行
	docker-compose up -d
	```
	**建议先在前台运行试用一下，看看有没有报错，如果启动和使用都没有问题，再改成后台运行。**

  启动成功之后，访问`http://你的机器ip`即可看到网页内容。

## 使用最新版本docker镜像启动

- 如果你只是想自己使用docker部署，可以直接使用我已经打包好的镜像和前端资源

- 首先将前端资源的压缩包解压

	前端资源的压缩包在`/docker-compose/nginx/html.zip`，使用`unzip html.zip`将其解压缩。你应该可以看到`/docker-compose/nginx/html`下面有`index.html`和其他前端资源。

- 然后给你的网站添加用户名和密码

	```shell
  # 进入文件夹 `/docker-compose/nginx`
  cd docker-compose/nginx
  # 运行add_user.sh脚本，根据提示创建用户名和密码
  # (密码文件将被保存在 /docker-compose/nginx/auth/.htpasswd)
  bash add_user.sh
  # 如果你想删除一个用户，可以使用remove_user.sh脚本
  bash remove_user.sh
	```


- 最后修改`docker-compose/docker-compose.yml`文件。

	除了填写你自己的`OPENAI_API_KEY`之外，还要根据自己的系统环境修改`image`的标签，如果你部署用的是x86架构的机器，就填写`wenjing95/chatgpt-web-backend:x86_64`，如果你用的是arm架构的机器，就填写`wenjing95/chatgpt-web-backend:aarch64`。

	```
	version: '3'

	services:
	  app:
      # 根据自己的系统选择x86_64还是aarch64
      image: wenjing95/chatgpt-web-backend:x86_64
      # image: wenjing95/chatgpt-web-backend:aarch64
      ports:
        - 3002:3002
      environment:
        # 记得填写你的OPENAI_API_KEY
        OPENAI_API_KEY: your_openai_api_key
        # 访问OpenAI的超时时间(毫秒)，可选，默认值为 '100000'
        OPENAI_TIMEOUT_MS: '100000'
        # 可选，默认值为 gpt-3.5-turbo
        API_MODEL: gpt-3.5-turbo
        # Socks代理，可选，格式为 http://127.0.0.1:10808
        SOCKS_PROXY: ''
        # HOST，可选，默认值为 0.0.0.0
        HOST: 0.0.0.0
        # PORT，可选，默认值为 3002
        PORT: 3002
    nginx:
      depends_on:
        - app
      image: nginx:alpine
      ports:
        - '80:80'
      expose:
        - '80'
      volumes:
        - ./nginx/html/:/etc/nginx/html/
        - ./nginx/nginx.conf:/etc/nginx/conf.d/default.conf
      links:
        - app
	```

- 最后进入文件夹 `/docker-compose` 运行以下命令

	```shell
	# 前台运行
	docker-compose up
	# 或后台运行
	docker-compose up -d
	```
	**建议先在前台运行试用一下，看看有没有报错，如果启动和使用都没有问题，再改成后台运行。**

	启动成功之后，访问`http://你的机器ip`即可看到网页内容。

## 常见问题

Q: 为什么 `Git` 提交总是报错？

A: 因为有提交信息验证，请遵循 [Commit 指南](./CONTRIBUTING.md)

Q: 如果只使用前端页面，在哪里改请求接口？

A: 根目录下 `.env` 文件中的 `VITE_GLOB_API_URL` 字段。

Q: 文件保存时全部爆红?

A: `vscode` 请安装项目推荐插件，或手动安装 `Eslint` 插件。

Q: 前端没有打字机效果？

A: 一种可能原因是经过 Nginx 反向代理，开启了 buffer，则 Nginx
会尝试从后端缓冲一定大小的数据再发送给浏览器。请尝试在反代参数后添加 `proxy_buffering off;`，然后重载 Nginx。其他 web
server 配置同理。

Q: 为什么录音功能不能用？

A: 录音需要https环境，推荐使用cloudflare的免费https证书。

Q: build docker容器的时候，显示`exec entrypoint.sh: no such file or directory`？

A: 因为`entrypoint.sh`文件的换行符是`LF`，而不是`CRLF`，如果你用`CRLF`的IDE操作过这个文件，可能就会出错。可以使用`dos2unix`工具将`LF`换成`CRLF`。

## 参与贡献

贡献之前请先阅读 [贡献指南](./CONTRIBUTING.md)

感谢原作者[Chanzhaoyu](https://github.com/Chanzhaoyu/chatgpt-web/)和所有做过贡献的人，还有生产力工具`ChatGpt`
和`Github Copilot`!

<a href="https://github.com/WenJing95/chatgpt-web/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=WenJing95/chatgpt-web" />
</a>

## 赞助

如果你觉得这个项目对你有帮助，请给我点个Star。

如果情况允许，请支持原作者[Chanzhaoyu](https://github.com/Chanzhaoyu/chatgpt-web/)

## License

MIT © [WenJing95](./license)


CHANGELOG.md
## v2.10.9

`2023-04-11`

## Feature
- 根据浏览器环境设置默认语言

`2023-04-03`

## Feature
- 增加了简单的权限验证

`2023-03-26`

## Fix
- 前端跟随主项目修复了bug和优化体验
- 后端修复了可能导致死循环的bug

## Feature
- 预备支持gpt-4
- 更新token计算方式

`2023-03-21`

## Fix
- 修复bug

## Feature
跟随主项目更新了以下内容
1. 聊天中的链接改成新窗口打开
2. AI回复的消息可以显示原文

`2023-03-14`

## Fix
- 优化一些逻辑
- 修复bug

`2023-03-12`

## Style
- 提高易用性

`2023-03-08`

## Feature
- 基于原版v2.9.3版本重新开始
- 可以识别语音消息：通过接入`whisper-1`API实现
- 可以微调参数：通过设置可以节省token消耗量、让`ChatGpt`更准确地回答问题
- `ChatGpt Web`现在可以“记住”更多的上下文：服务器端将聊天记录被持久化保存
- 去掉了`ChatGPTUnofficialProxyAPI`和反向代理功能：考虑到使用者并非都有精力折腾，因此只保留了官方API的方式。
- 可以随机生成头像
- 增加日语界面
- 优化了一点移动端体验
- 后端使用python重写

`2023-03-06`

## Enhancement
- 感谢 [ChandlerVer5](https://github.com/Chanzhaoyu/chatgpt-web/pull/305) 使用 `markdown-it` 替换 `marked`，解决代码块闪烁的问题
- 感谢 [shansing](https://github.com/Chanzhaoyu/chatgpt-web/pull/277) 改善文档
- 感谢 [nalf3in](https://github.com/Chanzhaoyu/chatgpt-web/pull/293) 添加英文翻译

## BugFix
- 感谢[sepcnt ](https://github.com/Chanzhaoyu/chatgpt-web/pull/279) 修复切换记录时编辑状态未关闭的问题
- 修复复制代码的兼容性报错问题
- 修复部分优化小问题

## v2.9.2

`2023-03-04`

手动部署的同学，务必删除根目录和`service`中的`node_modules`重新安装依赖，降低出现问题的概率，自动部署的不需要做改动。

### Feature
- 感谢 [hyln9](https://github.com/Chanzhaoyu/chatgpt-web/pull/247) 添加对渲染 `LaTex` 数学公式的支持
- 感谢 [ottocsb](https://github.com/Chanzhaoyu/chatgpt-web/pull/227) 添加支持 `webAPP` (苹果添加到主页书签访问)支持
- 添加 `OPENAI_API_BASE_URL` 可选环境变量[#249]
## Enhancement
- 优化在高分屏上主题内容的最大宽度[#257]
- 现在文字按单词截断[#215][#225]
### BugFix
- 修复动态生成时代码块不能被复制的问题[#251][#260]
- 修复 `iOS` 移动端输入框不会被键盘顶起的问题[#256]
- 修复控制台渲染警告
## Other
- 更新依赖至最新
- 修改 `README` 内容

## v2.9.1

`2023-03-02`

### Feature
- 代码块添加当前代码语言显示和复制功能[#197][#196]
- 完善多语言，现在可以切换中英文显示

## Enhancement
- 由[Zo3i](https://github.com/Chanzhaoyu/chatgpt-web/pull/187) 完善 `docker-compose` 部署文档

### BugFix
- 由 [ottocsb](https://github.com/Chanzhaoyu/chatgpt-web/pull/200) 修复头像修改不同步的问题
## Other
- 更新依赖至最新
- 修改 `README` 内容
## v2.9.0

`2023-03-02`

### Feature
- 现在能复制带格式的消息文本
- 新设计的设定页面，可以自定义姓名、描述、头像（链接方式）
- 新增`403`和`404`页面以便扩展

## Enhancement
- 更新 `chatgpt` 使 `ChatGPTAPI` 支持 `gpt-3.5-turbo-0301`（默认）
- 取消了前端超时限制设定

## v2.8.3

`2023-03-01`

### Feature
- 消息已输出内容不会因为中断而消失[#167]
- 添加复制消息按钮[#133]

### Other
- `README` 添加声明内容

## v2.8.2

`2023-02-28`
### Enhancement
- 代码主题调整为 `One Dark - light|dark` 适配深色模式
### BugFix
- 修复普通文本代码渲染和深色模式下的问题[#139][#154]

## v2.8.1

`2023-02-27`

### BugFix
- 修复 `API` 版本不是 `Markdown` 时，普通 `HTML` 代码会被渲染的问题 [#146]

## v2.8.0

`2023-02-27`

- 感谢 [puppywang](https://github.com/Chanzhaoyu/chatgpt-web/commit/628187f5c3348bda0d0518f90699a86525d19018) 修复了 `2.7.0` 版本中关于流输出数据的问题（使用 `nginx` 需要自行配置 `octet-stream` 相关内容）

- 关于为什么使用 `octet-stream` 而不是 `sse`，是因为更好的兼容之前的模式。

- 建议更新到此版本获得比较完整的体验

### Enhancement
- 优化了部份代码和类型提示
- 输入框添加换行提示
- 移动端输入框现在回车为换行，而不是直接提交
- 移动端双击标题返回顶部，箭头返回底部

### BugFix
- 流输出数据下的问题[#122]
- 修复了 `API Key` 下部份代码不换行的问题
- 修复移动端深色模式部份样式问题[#123][#126]
- 修复主题模式图标不一致的问题[#126]

## v2.7.3

`2023-02-25`

### Feature
- 适配系统深色模式 [#118](https://github.com/Chanzhaoyu/chatgpt-web/issues/103)
### BugFix
- 修复用户消息能被渲染为 `HTML` 问题 [#117](https://github.com/Chanzhaoyu/chatgpt-web/issues/117)

## v2.7.2

`2023-02-24`
### Enhancement
- 消息使用 [github-markdown-css](https://www.npmjs.com/package/github-markdown-css) 进行美化，现在支持全语法
- 移除测试无用函数

## v2.7.1

`2023-02-23`

因为消息流在 `accessToken` 中存在解析失败和消息不完整等一系列的问题，调整回正常消息形式

### Feature
- 现在可以中断请求过长没有答复的消息
- 现在可以删除单条消息
- 设置中显示当前版本信息

### BugFix
- 回退 `2.7.0` 的消息不稳定的问题

## v2.7.0

`2023-02-23`

### Feature
- 使用消息流返回信息，反应更迅速

### Enhancement
- 样式的一点小改动

## v2.6.2

`2023-02-22`
### BugFix
- 还原修改代理导致的异常问题

## v2.6.1

`2023-02-22`

### Feature
- 新增 `Railway` 部署模版

### BugFix
- 手动打包 `Proxy` 问题

## v2.6.0

`2023-02-21`
### Feature
- 新增对 `网页 accessToken` 调用 `ChatGPT`，更智能不过不太稳定 [#51](https://github.com/Chanzhaoyu/chatgpt-web/issues/51)
- 前端页面设置按钮显示查看当前后端服务配置

### Enhancement
- 新增 `TIMEOUT_MS` 环境变量设定后端超时时常（单位：毫秒）[#62](https://github.com/Chanzhaoyu/chatgpt-web/issues/62)

## v2.5.2

`2023-02-21`
### Feature
- 增加对 `markdown` 格式的支持 [Demo](https://github.com/Chanzhaoyu/chatgpt-web/pull/77)
### BugFix
- 重载会话时滚动条保持

## v2.5.1

`2023-02-21`

### Enhancement
- 调整路由模式为 `hash`
- 调整新增会话添加到
- 调整移动端样式


## v2.5.0

`2023-02-20`

### Feature
- 会话 `loading` 现在显示为光标动画
- 会话现在可以再次生成回复
- 会话异常可以再次进行请求
- 所有删除选项添加确认操作

### Enhancement
- 调整 `chat` 为路由页面而不是组件形式
- 更新依赖至最新
- 调整移动端体验

### BugFix
- 修复移动端左侧菜单显示不完整的问题

## v2.4.1

`2023-02-18`

### Enhancement
- 调整部份移动端上的样式
- 输入框支持换行

## v2.4.0

`2023-02-17`

### Feature
- 响应式支持移动端
### Enhancement
- 修改部份描述错误

## v2.3.3

`2023-02-16`

### Feature
- 添加 `README` 部份说明和贡献列表
- 添加 `docker` 镜像
- 添加 `GitHub Action` 自动化构建

### BugFix
- 回退依赖更新导致的 [Eslint 报错](https://github.com/eslint/eslint/issues/16896)

## v2.3.2

`2023-02-16`

### Enhancement
- 更新依赖至最新
- 优化部份内容

## v2.3.1

`2023-02-15`

### BugFix
- 修复多会话状态下一些意想不到的问题

## v2.3.0

`2023-02-15`
### Feature
- 代码类型信息高亮显示
- 支持 `node ^16` 版本
- 移动端响应式初步支持
- `vite` 中 `proxy` 代理

### Enhancement
- 调整超时处理范围

### BugFix
- 修复取消请求错误提示会添加到信息中
- 修复部份情况下提交请求不可用
- 修复侧边栏宽度变化闪烁的问题

## v2.2.0

`2023-02-14`
### Feature
- 会话和上下文本地储存
- 侧边栏本地储存

## v2.1.0

`2023-02-14`
### Enhancement
- 更新依赖至最新
- 联想功能移动至前端提交，后端只做转发

### BugFix
- 修复部份项目检测有关 `Bug`
- 修复清除上下文按钮失效

## v2.0.0

`2023-02-13`
### Refactor
重构并优化大部分内容

## v1.0.5

`2023-02-12`

### Enhancement
- 输入框焦点，连续提交

### BugFix
- 修复信息框样式问题
- 修复中文输入法提交问题

## v1.0.4

`2023-02-11`

### Feature
- 支持上下文联想

## v1.0.3

`2023-02-11`

### Enhancement
- 拆分 `service` 文件以便扩展
- 调整 `Eslint` 相关验证

### BugFix
- 修复部份控制台报错

## v1.0.2

`2023-02-10`

### BugFix
- 修复新增信息容器不会自动滚动到问题
- 修复文本过长不换行到问题 [#1](https://github.com/Chanzhaoyu/chatgpt-web/issues/1)


service/main.py
import openai
import os
from os.path import abspath, dirname
from loguru import logger
from chatgpt_wapper import process
from fastapi import FastAPI, UploadFile, File
from fastapi.responses import JSONResponse, StreamingResponse
import uvicorn
from message_store import MessageStore
from whisper_wapper import process_audio
import argparse
from api_model import ApiModel

log_folder = os.path.join(abspath(dirname(__file__)), "log")
logger.add(os.path.join(log_folder, "{time}.log"), level="INFO")

DEFAULT_TIMEOUT_MS_STRING = "100000"
MIN_TIMEOUT_MS = 15000
DEFAULT_DB_SIZE = 100000
DEFAULT_API_MODEL = "gpt-3.5-turbo"
DEFAULT_MAX_TOKEN = 4096

massage_store = MessageStore(db_path="message_store.json", table_name="chatgpt", max_size=DEFAULT_DB_SIZE)

app = FastAPI()

stream_response_headers = {
    "Content-Type": "application/octet-stream",
    "Cache-Control": "no-cache",
}


@app.post("/config")
async def config():
    return JSONResponse(content=dict(
        message=None,
        status="Success",
        data=dict(
            apiModel=API_MODEL,
            socksProxy=SOCKS_PROXY,
            timeoutMs=OPENAI_TIMEOUT * 1000,
        )
    ))


@app.post("/chat-process")
async def chat_process(request_data: dict):
    prompt = request_data["prompt"]
    options = request_data["options"]

    if 1 == request_data["memory"]:
        memory_count = 5
    elif 50 == request_data["memory"]:
        memory_count = 20
    else:
        memory_count = 999

    if 1 == request_data["top_p"]:
        top_p = 0.2
    elif 50 == request_data["top_p"]:
        top_p = 0.5
    else:
        top_p = 1

    answer_text = process(prompt, options, memory_count, top_p, MASSAGE_STORE, OPENAI_TIMEOUT, MAX_TOKEN,
                          model=API_MODEL)
    return StreamingResponse(content=answer_text, headers=stream_response_headers, media_type="text/event-stream")


@app.post("/audio-chat-process")
async def audio_chat_process(audio: UploadFile = File(...)):
    prompt = process_audio(audio, OPENAI_TIMEOUT, "whisper-1")
    return StreamingResponse(content=prompt, headers=stream_response_headers, media_type="text/event-stream")


def init_config():
    # 读取配置
    parser = argparse.ArgumentParser(description='')
    parser.add_argument('--openai_api_key', type=str, help='API key for OpenAI')
    parser.add_argument('--api_model', type=str, default="gpt-3.5-turbo",
                        help='OpenAI API model, default is gpt-3.5-turbo')
    parser.add_argument('--socks_proxy', type=str, default="",
                        help='Socks proxy, default is "", e.g. http://127.0.0.1:10808')
    parser.add_argument('--timeout_ms', type=str, default=DEFAULT_TIMEOUT_MS_STRING,
                        help="(Deprecate) Timeout for OpenAI API, default is '100000'")
    parser.add_argument('--openai_timeout_ms', type=str, default=DEFAULT_TIMEOUT_MS_STRING,
                        help="Timeout for OpenAI API, default is '100000'")
    # parser.add_argument('--service_timeout_ms', type=str, default=DEFAULT_TIMEOUT_MS_STRING,
    #                     help="Timeout for backend service, default is '100000'")
    parser.add_argument('--host', type=str, default="0.0.0.0", help='Host for server, default is 0.0.0.0')
    parser.add_argument('--port', type=str, default="3002", help="Port for server, default is '3002'")
    args = parser.parse_args()

    if not args.openai_api_key:
        err = "OpenAI API key is not found. use --openai_api_key to set it."
        logger.error(err)
        raise TypeError(err)
    openai_api_key = args.openai_api_key
    openai.api_key = args.openai_api_key

    input_api_model = args.api_model
    if not input_api_model:
        err = "API model is not found."
        logger.error(err)
        raise TypeError(err)

    if input_api_model not in ApiModel.KNOWN_API_MODEL_NAMES:
        warning = "Unknown Api model '{}'. Legal settings are {}, The system has automatically set it to {}".format(
            input_api_model,
            ApiModel.KNOWN_API_MODEL_NAMES,
            DEFAULT_API_MODEL
        )
        logger.warning(warning)

    api_model = ApiModel.get_api_model_name(input_api_model, DEFAULT_API_MODEL)
    max_token = ApiModel.get_max_token(api_model, DEFAULT_MAX_TOKEN)

    socks_proxy = args.socks_proxy
    if socks_proxy:
        logger.info("Socks proxy is enabled.")
        logger.info("Socks proxy is {}.".format(socks_proxy))
        openai.proxy = socks_proxy
    else:
        logger.info("Socks proxy is disabled.")

    if DEFAULT_TIMEOUT_MS_STRING != args.timeout_ms:
        logger.warning("The parameter '--timeout_ms' is deprecated, please use '--openai_timeout_ms' instead.")
        args.openai_timeout_ms = args.timeout_ms

    openai_timeout_ms = args.openai_timeout_ms or DEFAULT_TIMEOUT_MS_STRING

    if isinstance(openai_timeout_ms, str):
        try:
            openai_timeout_ms = int(openai_timeout_ms)
        except:
            openai_timeout_ms = DEFAULT_TIMEOUT_MS_STRING

    if openai_timeout_ms < MIN_TIMEOUT_MS:
        openai_timeout_ms = MIN_TIMEOUT_MS
        logger.warning(
            "OpenAI timeout is too short, the system has automatically set it to {openai_timeout_ms}(ms).".format(
                openai_timeout_ms=MIN_TIMEOUT_MS))

    openai_timeout = openai_timeout_ms / 1000

    # service_timeout_ms = args.service_timeout_ms or 100000
    # if isinstance(service_timeout_ms, str):
    #     try:
    #         service_timeout_ms = int(service_timeout_ms)
    #     except:
    #         service_timeout_ms = 100000

    # if service_timeout_ms < 15000:
    #     service_timeout_ms = 15000
    #     logger.warning("Service timeout is too short, the system has automatically set it to 15000(ms).")

    # service_timeout = service_timeout_ms / 1000

    # if openai_timeout_ms > service_timeout_ms:
    #     openai_timeout = service_timeout
    #     logger.warning(
    #         "OpenAI timeout is longer than service timeout, the system has automatically set it to the same as service timeout.")

    host = args.host or "0.0.0.0"
    port = args.port or 3002
    if isinstance(port, str):
        try:
            port = int(port)
        except:
            err = "Port must be a number."
            logger.error(err)
            raise TypeError(err)

    return massage_store, openai_api_key, host, port, api_model, max_token, socks_proxy, openai_timeout


if __name__ == "__main__":
    MASSAGE_STORE, OPENAI_API_KEY, HOST, PORT, API_MODEL, MAX_TOKEN, SOCKS_PROXY, OPENAI_TIMEOUT = init_config()
    logger.info("OPENAI_API_KEY:{}".format(OPENAI_API_KEY))
    logger.info("HOST:{}".format(HOST))
    logger.info("PORT:{}".format(PORT))
    logger.info("API_MODEL:{}".format(API_MODEL))
    logger.info("MAX_TOKEN:{}".format(MAX_TOKEN))
    logger.info("SOCKS_PROXY:{}".format(SOCKS_PROXY))
    logger.info("OPENAI_TIMEOUT_MS:{}".format(OPENAI_TIMEOUT * 1000))
    # logger.info("SERVICE_TIMEOUT_MS:{}".format(SERVICE_TIMEOUT * 1000))

    uvicorn.run(app, host=HOST, port=PORT)
